doc/man1/tigerxmlimport.1
.\" Man page for tigerxmlimport
.\" Use the following command to view man page:
.\"
.\"  tbl tigerxmlimport.1 | nroff -man | less
.\"
.TH TIGERXMLIMPORT 1 "May 25, 2008"
.SH NAME
tigerxmlimport \- A tool to convert TIGER XML format to Emdros MQL
.SH SYNOPSIS
\fBtigerxmlimport\fR [ options ] [input_filename ...]
.br
.SH DESCRIPTION
\fBtigerxmlimport\fR is a command-line tool for converting a corpus
stored as TIGER XML to Emdros MQL statements that create object types
and/or objects.  These can then later be imported into Emdros, e.g.,
with the mql(1) program.

Only the data definition part of TIGER XML is supported, not the parts
that represent queries or matches.  This is not really a bug, since
the purpose of the program is to import data, not queries or matches.


.SH OPTIONS
\fBtigerxmlimport\fR supports the following command-line switches:
.TP 6
.BI \-\-help
show help, then quit
.TP
.BI \-V\ ,\ \-\-version
show version, then quit
.TP
.BI \-b\ ,\ \-\-backend \ backend
set database backend to `backend'. Valid values are: For PostgreSQL:
"p", "pg", "postgres", and "postgresql". For MySQL: "m", "my", and
"mysql". For SQLite 2.X.X: "2", "s", "l", "lt", "sqlite", and
"sqlite2". For SQLite 3.X.X: "3", "s3", "lt3", and "sqlite3".
.TP
.BI \-s\ ,\ \-\-schema
print MQL schema to the output stream, before dumping data.
.TP
.BI \-d\ ,\ \-\-dbname \ dbname 
set database name. If used with -s, the string "CREATE DATABASE
'dbname' GO USE DATABASE 'dbname' GO" will be issued before the schema.  If used when importing, "USE DATABASE 'dbname' GO" will be issued before anything else.
.TP
.BI \-o\ ,\ \-\-output \ filename
dump to file filename. The default is "-", which means "standard
output".
.TP
.BI \-\-start-monad \ monad
The start monad to use. Must be >= 1.  The default is 1.
.TP
.BI \-\-start-id_d \ id_d
The start id_d to use. Must be >= 1.  The default is 1.
.TP
.BI \-\-ntotf \ feature\-name
The feature name to use in case you want to have separate object types
for NonTerminals.  Each value for the named feature will become an
object type, and the TigerXML feature will not show up as an EMdF
feature on the object type.  Renaming will occur if a value is not a
valid object type name (i.e., if it is not a C identifier).  The
renaming is detailed below.  If this switch is not given, all
NonTerminals are gathered into a single object type called
"Nonterminal".
.TP
.BI --termotn \ object\-type\-name
The object type name to use for terminals.  If this switch is not
given, the terminals will be placed into an object type named
"Terminal".

.SH OPERATION

\fBtigerxmlimport\fR reads corpus data in TIGER XML format, and
converts the data to MQL statements for later importing into Emdros.

The filename given after the options on the command line is opened,
and it must contain a TIGER XML document.  If no filename is given,
the input is read from stdin.

If no \fB-o\fR switch is given, the output is printed on stdout.

If an error occurs, the string "FAILURE" or the string "ERROR" is
printed on stderr, along with an error message.

If no error occurs, a string of the form "SUCCESS: next_monad is X
next_id_d is Y" is printed on stderr, where X and Y are positive
integers denoting the next monad and the next id_d to be used by the
next invocation of the program, respectively.  This is useful if
you've got several directories' worth of documents to import.

.SH SCHEMA

The schema can be seen by giving the program \fB-s\fR switch, with an
optional \fB-d\fR switch.

A "Sentence" corresponds to one top-level sentence.  It has a single
feature, called "id", giving the corpus id of the sentence.  The "id"
feature is of type STRING.

If the --ntotf switch is not given, all nonterminals are gathered into
a single object type, called "Nonterminal".  A "Nonterminal" is a
nonterminal in the tree which is not a top-level sentence.  Note that
this encompasses clauses ("S") as well as phrases.  Its "parent"
feature is an id_d that points to its immediate parent.  Its "edge"
feature is a STRING which shows the edge label <annotation> tag held.

The rest of the features of a "Nonterminal" object are computed as
follows: Inside the <annotation> tag, each <feature> tag whose domain
is either "NT" or "FREC" becomes an Emdros feature of type STRING FROM
SET, with the name given in the "name" attribute of the <feature> tag.
For example, <feature name="cat" domain="NT"> will become one Emdros
feature, with the name "cat" and the type "STRING FROM SET".

Secondary edges are given as secedgeX (a STRING FROM SET) and
secedgeparentX (an id_d pointing to the secondary edge parent), where
X is an integer (1,2,3,4.... etc.).  The number of pairs of
secedgeX/secedgeparentX features depends on the maximum number of
secondary edges actually found in the corpus imported.  For example,
if there are a maximum of two secondary edges on any node in the
corpus, then the features will be secedge1, secedgeparent1, secedge2,
secedgeparent2.

If the --ntotf switch is given, the same as the above occurs, except
that the TigerXML feature named by the --ntotf switch will not be
added to any nonterminal object.  Instead, the feature named by the
switch will be interpreted as an object type name, and all
nonterminals with the same value for this feature will be placed into
an object type of the same name.  For example, if the switch + value
"--ntotf cat" are given, then the TigerXML feature "cat" is used.  If
a Nonterminal has a "cat" of "CL", then that nonterminal will be
placed into the object type named "CL".  If the value is not a valid
object type name (i.e., if it is not a valid C identifier), then the
value is renamed in order to become a valid object type name.  This is
done as follows: All characters which do not belong to the set
[A-Za-z0-9_] are first made into underscores.  Then, if the result
starts with a digit ("[0-9]"), then an underscore is prepended.  Then,
if the result is either a reserved MQL word, or if the name already
exists in the set of object types to be used, underscores are appended
until the name is unique and not an MQL reserved word.

If the "--termotn" switch is not given, all terminals are gathered
into an object type by the name "Terminal".  A "Terminal" is a
terminal node in the tree (i.e., a token). Its features are calculated
precisely as for Nonterminal, except that the domain of the <feature>
tag is either "T" or "FREC", but not "NT".

If the "--termotn" switch, on the other hand, is given, then all
terminals are gathered into an object type with the name given by the
switch.


Since the schema is created from the <feature> parts of the header,
the program needs to read the corpus before it can emit the schema.
If you give the -s switch to the program, the schema will be printed
on stdout.


.SH EXAMPLES

tigerxmlimport -o data.mql -d mycorpus --schema \\
corpus.xml > schema.mql

This example reads the corpus.xml corpus (which must be in TIGER XML
format), and writes the data to the data.mql file, while writing the
schema to schema.mql.  The Emdros database will be called 'mycorpus'

Afterwards, the following will create the database:

mql schema.mql

mql data.mql




.SH BUGS

The subcorpus feature of TIGER XML is not supported.


.SH RETURN VALUES
.TP
.BR 0 " Success"
.TP
.BR 1 " Wrong usage"
.TP
.BR 2 " Connection to backend server could not be established"
.TP
.BR 3 " An exception occurred (the type is printed on stderr)"
.TP
.BR 4 " Could not open file"
.TP
.BR 5 " Database error"
.TP
.BR 6 " Compiler error (internal error)"
.SH AUTHORS
Written Ulrik Sandborg-Petersen (ulrikp@emdros.org).


NEWS
This file is for release-notes for Emdros.

- *** Version 3.7.0 *** 

Release notes for Emdros 3.7.0
==============================

- Emdros has been relicensed under the MIT license.

- All copyright in Emdros and related documentation and build scripts
  has been assigned to Ulrik Sandborg-Petersen's holding company,
  Sandborg-Petersen Holding ApS.  This is so as to be able to serve
  customers better with support and other Emdros-related help.

- A new construct was added to the topographic part of MQL: computed
  features, which take a monad set and returns a monad or integer:

  - first_monad(<monad_set_name>) returns the smallest (first) monad.
  - last_monad(<monad_set_name>) returns the largest (last) monad.
  - monad_count(<monad_set_mame>) returns the cardinality.
  - monad_set_length(<monad_set_mame>) returns last_monad - first_monad + 1.

  If the monad_set_name (and the parentheses) are missing, then the
  privileged monad set "monads" is assumed.

  For example:

  SELECT ALL OBJECTS
  WHERE
  [Clause
     monad_set_length > 15 AND monad_count <= 15
  ]

  This also works with object references:

  // Finds sentences and clauses which are co-extensive.
  SELECT ALL OBJECTS
  WHERE
  [Sentence AS s1
     [Clause monad_count = s1.monad_count]
  ]

- The following SWIG bindings have been removed:

  - PHP 5
  - Ruby
  - Perl
  - Android

  If any user needs one of these back, please let Ulrik know.

  These SWIG bindings were removed because:

  - The language was EOL'ed (PHP5)
  - The Emdros SWIG bindings had not been tested for several years (Ruby, Perl)
  - They were redundant (Android).

  Users of Android should simply use the Java SWIG bindings instead.

- The Emdros Query Tool no longer supports TECkit.  TECkit support was
  removed because we now live in a Unicode world, and TECkit was
  supposed to convert from legacy 8-bit encodings to Unicode.

- The examples/ subdirectory is gone, making for a smaller tarball.
  The examples have been moved to the example-programs-emdros and
  example-databases GitHub repos. This affects:

  - The Emdros Chunking Tool
  - The HAL implementation (Hyperspace Analogue to Language of Burgess et al.)
  - The King James Version, Parsed.
  - The sfm example.

- The agexport(1) program is gone, since it had not been maintained or
  tested for many years.

- The build scripts have been totally reimagined.  In particular:

  - The m4/ directory has been totally reorganized.

  - The doc/ subdirectory has been totally revamped, with much
    documentation having been moved to the docs-emdros repo on GitHub.

  - The ./configure script now takes a mutually exclusive pair of
    switches: --enable-devel | --disable-devel switch.

    The default is --disable-devel.

    If --disable-devel is true, then the Emdros libraries,
    libpcre_emdros.*, libemdf.*, libmql.*, libutil.*, and
    libharvest.*.  Instead, the amalgamation is used in lieu of these.
    In addition, these libraries are not compiled, thus providing a
    much faster compile experience.  In addition, the Emdros programs
    are linked against the amalgamation library.

    If --enable-devel is true, then everything is compiled, including
    the amalgamation, but the programs are linked against the
    development libraries, not the amalgamation.  This is useful if
    you are developing Emdros.


- *** Version 3.6.1 *** 

Release notes for Emdros 3.6.1
==============================

Emdros version 3.6.1 is a maintenance release with the following
summary of changes since 3.6.0:

- The MQL query language has been extended with the ability to compare
  features with type LIST OF INTEGER, LIST OF ID_D, and LIST OF
  <enum-type> with an explicit list (including the empty list) using
  the "=" comparison operator.

  For example:

  SELECT ALL OBJECTS
  WHERE
  [Clause_atom
     daughters = () // Test for empty list
     OR
     daughters = (12345,67891) // Test for explicit, non-empty list
  ]
  GO

- Detection of PHP 5 and PHP 7 on macOS now works correctly.

- Compileability on Solaris 10 with Sun's CC C++ compiler has been
  restored.

- Compatibility with wxWidget 2.8.12 has been restored.


Enjoy!


Ulrik Sandborg-Petersen



- *** Version 3.6.0 *** 

Release notes for Emdros 3.6.0
==============================

The release notes below cover some of the important changes in Emdros
since 3.5.0.

- SQLite2 support has been removed, as version 2 of SQLite has not
  been updated by the SQLite authors for many, many years.  Also,
  SQLite2 support in Emdros has been been deprecated for some time.

- The MQL subsystem has a better query planner, which is able to
  choose the "Innermost First" strategy, rather than the default
  "Outermost First" strategy, under many more circumstances.

- The Emdros MySQL backend now has support for transactions, if
  available in the MySQL server.

- Assorted performance improvements to the following subsystems:

  - XML parsing (renderobjects, for example)
  - MQL

- Node.JS support has been removed.  The reason is that the latest
  SWIG version (3.0.12) does not support the latest Node.  Also, the
  Node.JS bindings never really worked.

- PCRE has been upgraded.

- SQLite3 has been upgraded.

- JSON output now escapes string characters correctly.

- Compiling Linux .rpm packages works on recent systems (CentOS 7,
  Fedora 28).

- Compiling Linux .deb packages works on recent systems (Ubuntu 16.04,
  Ubuntu 18.04, Debian 9.4).

- Building on Windows supports the latest Visual Studio.

- Lots of compiler warnings have been removed.

- the harvest template language has new constructs:

  - {{ inccounter <string> var <string> }}
  - {{ deccounter <string> var <string> }}

- The Full Text Search FTS3 subsystem is now able to obey

src/tigerxmlimport.cpp
/*
 * tigerxmlimport.cpp
 *
 * A front-end to the TIGER XML import
 * Created: 5/1-2001 (1st of May, 2001)
 * Last update: 5/11-2018
 *
 * Return codes:
 * 
 * 0 : Success
 * 1 : Wrong usage
 * 2 : Connection to backend server could not be established
 * 3 : An exception occurred
 * 4 : Could not open file
 * 5 : Database error
 * 6 : Compiler error
 *
 */

/*
 * Copyright (C) 2001-2018     Ulrik Sandborg-Petersen
 * Copyright (C) 2018-present  Sandborg-Petersen Holding ApS, Denmark
 *
 * Licensed under the MIT License.
 *
 * Please see the file COPYING in the root of the sources for more details.
 *
 */



#include <emdros-config.h>
// Do this because emdros-config might #define malloc,
// in which case some systems will fail to use the real 
// malloc, thus causing build errors.
#undef malloc

#include <iostream>
#include <sstream>
#include <fstream>
#include <string>
#include <emdros-lconfig.h>
#include <emdfdb.h>

#include <tigerxmlimporter.h>

#include <opt.h>




int exec_list(const std::list<std::string>& input_filename_list, std::ostream* pOutStream, const std::string& dbname, monad_m start_monad, id_d_t start_id_d, bool bPrintSchema, const std::string& nonterminal_object_type_feature, const std::string& terminal_object_type_name)
{
	int result = 0;

	TIGERXMLImporter *pImporter = new TIGERXMLImporter(start_monad, start_id_d,
							   nonterminal_object_type_feature,
							   terminal_object_type_name);

	bool bResult = false;
	if (input_filename_list.empty()) {
		bResult = pImporter->readDocumentFromStream(&std::cin);
	} else {
		bResult = pImporter->readDocumentsFromFiles(input_filename_list);
	}

	if (!bResult) {
		std::cerr << "FAILURE: Error occurred reading documents." << std::endl;
		result = 5;
	} else {
		if (bPrintSchema) {
			if (dbname != "") {
				(*pOutStream) << "CREATE DATABASE '" << dbname << "' GO\n\n"
					      << "USE DATABASE '" << dbname << "' GO\n\n";
			}
			
			pImporter->putSchema(pOutStream);
		}
		

		if (dbname != "") {
			(*pOutStream) << "USE DATABASE '" << dbname << "' GO\n\n";
		}

		pImporter->putCorpusAsMQL(pOutStream);

		std::cerr << "SUCCESS: next_monad is " << pImporter->getNextMonad() << "   next_id_d is " << pImporter->getNextID_D() << std::endl;
	}
	
	delete pImporter;
	
	return result;
}

bool open_file_for_writing(std::ofstream& fout, const std::string& filename)
{
	bool bResult = false;
	fout.open(filename.c_str());
	if (!fout) {
		std::cerr << "FAILURE: Could not open file " << filename << " for writing." << std::endl;
		std::cerr << "Action aborted." << std::endl;
		bResult = false;
	} else {
		bResult = true;
	}

	return bResult;
}

bool open_file_for_reading(std::ifstream& fin, const std::string& filename)
{
	bool bResult = false;
	fin.open(filename.c_str());
	if (!fin) {
		std::cerr << "FAILURE: Could not open file " << filename << " for reading." << std::endl;
		std::cerr << "Action aborted." << std::endl;
		bResult = false;
	} else {
		bResult = true;
	}

	return bResult;
}

void show_version(std::ostream& ostr)
{
	ostr << "tigerxmlimport from Emdros version " << EMDROS_VERSION << std::endl;
}

void print_usage(std::ostream& ostr)
{
	show_version(ostr);
	ostr << "Usage:" << std::endl;
	ostr << "   tigerxmlimport [options] [inputfile]" << std::endl;
	ostr << "   tigerxmlimport (--help|-V|--version)" << std::endl;
	ostr << "OPTIONS:" << std::endl;
	ostr << "   --help               Show this help" << std::endl;
	ostr << "   -V , --version       Show version" << std::endl;
	ostr << "   -s , --schema        Print schema" << std::endl;
	ostr << "   -d , --dbname db     database name (to use in MQL output)" << std::endl;
	ostr << "   -o , --output file   Dump database to filename file (\"-\" means stdout)" << std::endl;
	ostr << "   --start-monad monad  Start monad (default 1)" << std::endl;
	ostr << "   --start-id_d id_d    Start id_d (default 1)" << std::endl;
	ostr << "DEFAULTS:" << std::endl;
	ostr << "   -o -                 (Use stdout)" << std::endl;
	ostr << "   --start-monad 1" << std::endl;
	ostr << "   --start-id_d  1" << std::endl;
	ostr << "   If no filenames are given, a document is read from stdin." << std::endl;
	ostr << "   By default, no -d switch is assumed." << std::endl;
}

int main(int argc, char* argv[])
{
	// Set defaults
	std::string db;
	std::string output_filename("-");
	std::string terminal_object_type_name;
	std::string nonterminal_object_type_feature;
	std::list<std::string> input_filename_list;
	bool bPrintSchema = false;
	monad_m start_monad = 1;
	id_d_t start_id_d = 1;
	std::string dummy_hostname, dummy_user, dummy_password;
	eCharsets dummy_charset;
	eBackendKind dummy_backend_kind;

	addStandardArguments(false // Do not add -e | --encoding
			     );

	addOption("-s", "--schema");

	addOption("--ntotf", "--ntotf", true,
		  "", // Default value
		  "ERROR: --ntotf must have a feature name as its argument.\n"
		  "       example: --ntotf cat\n");

	addOption("--termotn", "--termotn", true,
		  "", // Default value
		  "ERROR: --termotn must have an object type as its argument.\n"
		  "       example: --termotn Token\n");

	addOption("-d", "--dbname", true, 
		  "", // Default value
		  "ERROR: -d and --dbname must have a database-name as their argument.\n"
		  "       example: -d book1\n");

	addOption("-o", "--output", true, 
		  "-", // Default value
		  "ERROR: -o and --output must have a filename (or \"-\" for stdout)\n"
		  "       as their argument.\n"
		  "       example: -o output.xml\n");


	addOption("--start-monad", "--start-monad", true, 
		  "1", // Default value
		  "ERROR: --start-monad must have a monad as its argument\n"
		  "       example: --start-monad 1\n");

	addOption("--start-id_d", "--start-id_d", true, 
		  "1", // Default value
		  "ERROR: --start-id_d must have an id_d as its argument\n"
		  "       example: --start-id_d 1\n");


	// Parse arguments
	std::list<std::string> surplus_arguments;

	std::string error_message;
	if (!parseArguments(argc, argv, error_message, surplus_arguments)) {
		print_usage(std::cerr);
		std::cerr << error_message << std::endl;
		return 1;
	} else {
		bool bShowVersion;
		bool bShowHelp;
		std::string error_message;
		if (!getStandardArguments(bShowVersion, bShowHelp,
					  dummy_hostname,
					  dummy_user,
					  dummy_password,
					  dummy_backend_kind,
					  dummy_charset,
					  error_message)) {
			print_usage(std::cerr);
			std::cerr << error_message << std::endl;
			return 1;
		}

		if (bShowVersion) {
			show_version(std::cout);
			return 0;
		}

		// --help
		if (bShowHelp) {
			print_usage(std::cout);
			return 0;
		}

		// --start-monad
		std::string start_string;
		if (getArgumentValue("--start-monad", start_string)) {
			if (string_is_number(start_string)) {
				start_monad = string2long(start_string);
			} else {
				print_usage(std::cerr);
				std::cerr << "ERROR: The argument of --start-monad must be a number, and not '" << start_string << "'." << std::endl;
				return 1;
			}
		}
		if (start_monad < 1) {
			print_usage(std::cerr);
			std::cerr << "ERROR: The argument of --start-monad must be >= 1." << std::endl;
			return 1;
		}

		// --start-id_d
		if (getArgumentValue("--start-id_d", start_string)) {
			if (string_is_number(start_string)) {
				start_id_d = string2long(start_string);
			} else {
				print_usage(std::cerr);
				std::cerr << "ERROR: The argument of --start-id_d must be a number, and not '" << start_string << "'." << std::endl;
				return 1;
			}
		}
		if (start_id_d < 1) {
			print_usage(std::cerr);
			std::cerr << "ERROR: The argument of --start-id_d must be >= 1." << std::endl;
			return 1;
		}

		// --dbname | -d
		getArgumentValue("-d", db);

		// --ntotf nonterminal-object-type-feature
		getArgumentValue("--ntotf", nonterminal_object_type_feature);

		// --termotn terminal-object-type-name
		getArgumentValue("--termotn", terminal_object_type_name);

		// Output
		getArgumentValue("-o", output_filename);

		// -s | --schema
		if (getArgumentPresent("-s")) {
			bPrintSchema = true;
		}

		// Any and all surplus arguments are input filenames...
		input_filename_list = surplus_arguments;

		if (input_filename_list.size() >= 2) {
			print_usage(std::cerr);
			std::cerr << "ERROR: You can have at most one filename on the command line." << std::endl;
			return 1;
		}
	}


	int nResult = 3;
	bool bContinue = true;
	try {
		std::ostream *pOut = 0;
		std::ofstream fout;
		if (output_filename == "-") {
			pOut = &std::cout;
		} else {
			if (!open_file_for_writing(fout, output_filename)) {
				nResult = 4;
				bContinue = false;
			} else {
				pOut = &fout;
			}
		}

		if (bContinue) {
			nResult = exec_list(input_filename_list, pOut, db, start_monad, start_id_d, bPrintSchema, nonterminal_object_type_feature, terminal_object_type_name);
		}
	} catch (EMdFNULLValueException& e) {
		std::cerr << "ERROR: EMdFNULLValueException (Database error)..." << std::endl;
		/*
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		*/
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFDBDBError& e) {
		std::cerr << "ERROR: EMdFDBDBError (Database error)..." << std::endl;
		/*
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		*/
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFDBException& e) {
		std::cerr << "ERROR: EMdFDBException (Database error)..." << std::endl;
		/*
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		*/
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableColumnException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableRowException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (BadMonadsException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (WrongCharacterSetException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFOutputException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EmdrosException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (...) {
		std::cerr << "Unknown exception occurred.  Program aborted." << std::endl;
	}

	// Clean up
	//	delete pEE;
	return nResult;
}


